java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java

```diff
@@ -0,0 +1,415 @@
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
+      if (leftOp.equals(rightOp)) {
+        if (shouldNotInverse().equals(booleanConstraint)) {
+          return ImmutableList.of(programState);
+        }
+        return ImmutableList.of();
+      }
+      List<ProgramState> result = copyConstraint(leftOp, rightOp, programState, booleanConstraint);
+      if (result.isEmpty()) {
+        return ImmutableList.of();
+      }
+      List<ProgramState> reversedResult = new ArrayList<>();
+      for (ProgramState ps : result) {
+        List<ProgramState> reversedStates = copyConstraint(rightOp, leftOp, ps, booleanConstraint);
+        if (reversedStates.size() == 1 && reversedStates.get(0) == programState) {
+          Map<SymbolicValue, Object> newConstraints = Maps.newHashMap(programState.constraints);
+          newConstraints.put(this, booleanConstraint);
+          reversedResult.add(new ProgramState(programState.values, newConstraints, programState.visitedPoints, programState.stack));
+        } else {
+          reversedResult.addAll(reversedStates);
+        }
+      }
+      return reversedResult;
+
+    }
+
+    @Override
+    public String toString() {
+      return "EQ_TO_" + super.toString();
+    }
+
+    private List<ProgramState> copyConstraint(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
+      Object constraintLeft = programState.constraints.get(from);
+      if (constraintLeft instanceof BooleanConstraint) {
+        BooleanConstraint boolConstraint = (BooleanConstraint) constraintLeft;
+        return to.setConstraint(programState, shouldNotInverse().equals(booleanConstraint) ? boolConstraint : boolConstraint.inverse());
+      } else if (constraintLeft instanceof NullConstraint) {
+        NullConstraint nullConstraint = (NullConstraint) constraintLeft;
+        if (nullConstraint.equals(NullConstraint.NULL)) {
+          return to.setConstraint(programState, shouldNotInverse().equals(booleanConstraint) ? nullConstraint : nullConstraint.inverse());
+        } else if (shouldNotInverse().equals(booleanConstraint)) {
+          return to.setConstraint(programState, nullConstraint);
+        }
+      }
+      return ImmutableList.of(programState);
+    }
+
+  }
+  class NotEqualToSymbolicValue extends BinarySymbolicValue {
+
+    public NotEqualToSymbolicValue(int id) {
+      super(id);
+    }
+
+    @Override
+    public String toString() {
+      return "NEQ_TO_" + super.toString();
+    }
+
+    @Override
+    BooleanConstraint shouldNotInverse() {
+      return BooleanConstraint.FALSE;
+    }
+  }
+  class EqualToSymbolicValue extends BinarySymbolicValue {
+
+    public EqualToSymbolicValue(int id) {
+      super(id);
+    }
+
+    @Override
+    BooleanConstraint shouldNotInverse() {
+      return BooleanConstraint.TRUE;
+    }
+
+  }
+
+  abstract class UnarySymbolicValue extends ObjectSymbolicValue {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Override this superclass' "equals" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2160)